RecoParticleFlow/PFProducer/plugins/MLPFProducerSonic.cc

```diff
@@ -0,0 +1,199 @@
+    assert(tensor_size <= NUM_MAX_ELEMENTS_BATCH);
+
+    auto& input1 = iInput.at("x");
+
+    //we ignore Sonic/Triton batching, as it doesn't create a dim-3 input for batch size 1.
+    //instead, we specify the batch dim as a model dim.
+    input1.setShape(0, 1);
+    input1.setShape(1, tensor_size);
+
+    auto data1 = std::make_shared<TritonInput<float>>(1);
+    auto& vdata1 = (*data1)[0];
+    vdata1.reserve(input1.sizeShape());
+
+    //Fill the input tensor
+    unsigned int ielem = 0;
+    for (const auto* pelem : all_elements) {
+      //get the PFElement as a pointer
+      const auto& elem = *pelem;
+
+      //prepare the input array from the PFElement
+      const auto& props = get_element_properties(elem);
+
+      //copy features to the input array
+      for (unsigned int iprop = 0; iprop < NUM_ELEMENT_FEATURES; iprop++) {
+        vdata1.push_back(normalize(props[iprop]));
+      }
+      ielem += 1;
+    }
+    input1.toServer(data1);
+  }
+
+  void produce(edm::Event& iEvent, edm::EventSetup const& iSetup, Output const& iOutput) override {
+    using namespace reco::mlpf;
+
+    std::vector<reco::PFCandidate> pOutputCandidateCollection;
+    const auto& output1 = iOutput.at("Identity");
+
+    //get the data of the first (and only) batch
+    const auto& out_data = output1.template fromServer<float>();
+
+    const auto num_elem = output1.shape()[1];
+
+    unsigned int num_pred_particles = 0;
+    for (unsigned int ielem = 0; ielem < num_elem; ielem++) {
+      //get the coefficients in the output corresponding to the class probabilities (raw logits)
+      std::vector<float> pred_id_logits;
+      for (unsigned int idx_id = 0; idx_id <= NUM_CLASS; idx_id++) {
+        pred_id_logits.push_back(out_data[0][ielem * NUM_OUTPUTS + idx_id]);
+      }
+
+      //get the most probable class PDGID
+      int pred_pid = pdgid_encoding[arg_max(pred_id_logits)];
+
+      //get the predicted momentum components
+      float pred_eta = out_data[0][ielem * NUM_OUTPUTS + IDX_ETA];
+      float pred_phi = out_data[0][ielem * NUM_OUTPUTS + IDX_PHI];
+
+      //put phi into a modular range
+      pred_phi = TMath::ATan2(TMath::Sin(pred_phi), TMath::Cos(pred_phi));
+      float pred_e = out_data[0][ielem * NUM_OUTPUTS + IDX_ENERGY];
+
+      //currently, set the pT from a massless approximation.
+      //later versions of the model may predict predict both the energy and pT of the particle
+      float pred_pt = pred_e / cosh(pred_eta);
+      float pred_charge = out_data[0][ielem * NUM_OUTPUTS + IDX_CHARGE];
+
+      //a particle was predicted for this PFElement
+      if (pred_pid != 0) {
+        //set the charge to +1 or -1 for PFCandidates that are charged, according to the sign of the predicted charge
+        reco::PFCandidate::Charge charge = 0;
+        if (pred_pid == 11 || pred_pid == 13 || pred_pid == 211) {
+          charge = pred_charge > 0 ? +1 : -1;
+        }
+
+        TLorentzVector p4;
```

Comment:
`TLorentzVector` is unnecessary in principle: probably better to use `PtEtaPhiELorentzVector` (https://github.com/cms-sw/cmssw/blob/master/DataFormats/Math/interface/LorentzVector.h)